### PR TITLE
Add math constants and logic functions

### PR DESCRIPTION
**Pull request by joaosferreira**
I've added the [math constants](https://numpy.org/devdocs/reference/constants.html) and [logic functions](https://numpy.org/doc/stable/reference/routines.logic.html) that were missing. To name  them separately:

Math constants:
- `euler_gamma`
- `nan` and aliases 
- `inf` and aliases
- `NINF`
- `PZERO`
- `NZERO`
- `newaxis`

Logic functions:
- `isposinf`
- `isneginf`
- `iscomplex`
- `iscomplexobj`
- `isreal`
- `isrealobj`
- `isscalar`
- `array_equal `
- `array_equiv `

Most of these additions aren't complete and some might be incorrect so this PR is a draft intended for me to get some advice. I'll give some thoughts on these:
1. I abandoned the idea of using singleton classes for the constants as they didn't seem the best approach, the result is very simple and follows the format of previously added constants.
2. The available argument replacers don't seem appropriate for `isposinf` and `isneginf`. I'm only marking the first argument as `Dispatchable` but looking at other multimethods it seems `out` should also be marked. So a custom argument replacer that handles only these two arguments seems to be needed here. Also, when should `out` be marked as non-coercible?
3. It seems that functions like `iscomplexobj` can receive any type of argument but I assume I should still mark it as an `ndarray` to be dispatched.
4. Both `array_equal` and `array_equiv` weren't foreseen in my proposal (probably overlooked them) but they should be added as well. I'm thinking of adding some simple defaults that do shape checking and use `allclose` .

### CODE REVIEW

**Comment by hameerabbasi** [comment]
> 1\. I abandoned the idea of using singleton classes for the constants as they didn't seem the best approach, the result is very simple and follows the format of previously added constants.

That seems okay to me.



> 2\. The available argument replacers don't seem appropriate for `isposinf` and `isneginf`. I'm only marking the first argument as `Dispatchable` but looking at other multimethods it seems `out` should also be marked. So a custom argument replacer that handles only these two arguments seems to be needed here. Also, when should `out` be marked as non-coercible?

A custom argument replacer is needed, or you can re-use the one from `ufunc.call`. Any arguments that can/will be mutated need to be marked as non-coercible, `out` is always one of those.



> 3\. It seems that functions like `iscomplexobj` can receive any type of argument but I assume I should still mark it as an `ndarray` to be dispatched.

Nope, setting a backend will dispatch it always. It also accepts dtypes and Python objects and doesn't coerce them, so I really don't think it should be marked as `ndarray`.



> 4\. Both `array_equal` and `array_equiv` weren't foreseen in my proposal (probably overlooked them) but they should be added as well. I'm thinking of adding some simple defaults that do shape checking and use `allclose` .

`allclose` itself needs a default, `array_equal` is simply `shape(a) == shape(b) and all(a == b)`, essentially.

**Comment by joaosferreira** [comment]
> A custom argument replacer is needed, or you can re-use the one from `ufunc.call`. Any arguments that can/will be mutated need to be marked as non-coercible, `out` is always one of those.

This seems simple enough:
```python
def _self_out_argreplacer(args, kwargs, dispatchables):
    def replacer(a, *args, out=None, **kwargs):
        return (dispatchables[0],), {"out": dispatchables[1]}

    return replacer(*args, **kwargs)
```
> Nope, setting a backend will dispatch it always. It also accepts dtypes and Python objects and doesn't coerce them, so I really don't think it should be marked as `ndarray`.

From what I understand I should remove `@all_of_type(ndarray)` and use `Dispatchable` to mark the argument but with which type?

> `allclose` itself needs a default, `array_equal` is simply `shape(a) == shape(b) and all(a == b)`, essentially.

You seem to have implemented `allclose_default`, right? Do you mean that it's incomplete?

The default for `array_equiv` is a little more complex, maybe this?

```python
def array_equiv_default(a1, a2):
    smaller, bigger = asarray(a1), asarray(a2)
    if ndim(a1) >= ndim(a2):
        smaller, bigger = a2, a1

    array_b = broadcast_to(smaller, bigger.shape)

    return all(array_b == bigger)
```

I also forgot to mention two things:
1. I didn't include `isfortran` as it says in the [docs](https://numpy.org/doc/stable/reference/generated/numpy.isfortran.html) that it's obsolete.
2. `isinf` is implemented as an ufunc, like this `isinf = ufunc("greater_equal", 1, 1)`, which I don't quite understand.

**Comment by hameerabbasi** [comment]
> 2\. `isinf` is implemented as an ufunc, like this `isinf = ufunc("greater_equal", 1, 1)`, which I don't quite understand.

That one's a bit tricky. `ufunc` objects aren't functions, they're custom objects with `__call__` methods. So we use those. Part of that is a bug, that should be `isinf = ufunc("isinf", 1, 1)` which basically means "`isinf` is a ufunc named `"isinf"`, with one input and one output."

**Comment by hameerabbasi** [comment]
> 1. I didn't include `isfortran` as it says in the [docs](https://numpy.org/doc/stable/reference/generated/numpy.isfortran.html) that it's obsolete.

That's a good idea.



> You seem to have implemented `allclose_default`, right? Do you mean that it's incomplete?

In that case, it can probably be used.



> The default for `array_equiv` is a little more complex, maybe this?

Nope, just `all(a1 == a2)` will work. All binary operations in NumPy broadcast automatically. 😄 

```ipython
>>> import numpy as np
>>> a = np.arange(5)
>>> b = np.arange(5)[:, None]
>>> a.shape
(5,)
>>> b.shape
(5, 1)
>>> np.all(a == b)
False
>>> (a == b).shape
(5, 5)
```

File: unumpy/_multimethods.py [hunk 1]
```diff
@@ -436,6 +444,48 @@ def all(a, axis=None, out=None, keepdims=False):
...
+@create_numpy(_self_argreplacer)
+@all_of_type(ndarray)
+def iscomplex(x):
+    return (x,)
+
+
+@create_numpy(_self_argreplacer)
+@all_of_type(ndarray)
+def iscomplexobj(x):
+    return (x,)
```
Comment (hameerabbasi) [inline unumpy/_multimethods.py]:
```suggestion
@create_numpy(_identity_argreplacer)
@all_of_type(ndarray)
def iscomplexobj(x):
    return ()
```

Comment (joaosferreira) [reply unumpy/_multimethods.py]: Isn't `all_of_type` unnecessary then?

Comment (hameerabbasi) [reply unumpy/_multimethods.py]: Yes, it's unnecessary.

File: unumpy/_multimethods.py [hunk 2]
```diff
@@ -436,6 +444,54 @@ def all(a, axis=None, out=None, keepdims=False):
     return (a, mark_non_coercible(out))
 
 
+def _self_out_argreplacer(args, kwargs, dispatchables):
+    def replacer(a, *args, out=None, **kwargs):
+        return (dispatchables[0],), {"out": dispatchables[1]}
```
Comment (hameerabbasi) [inline unumpy/_multimethods.py]:
```suggestion
        return (dispatchables[0],) + args, {"out": dispatchables[1], **kwargs}
```

Comment (hameerabbasi) [reply unumpy/_multimethods.py]: You must return the complete set of args/kwargs, not only the relevant ones. :)

**Comment by hameerabbasi** [comment]
CI is failing, you'll need to add an exception in the tests file for the XND backend.

**Comment by hameerabbasi** [comment]
Some merge conflicts to fix and we should be ready to go. 😄

**Comment by joaosferreira** [comment]
This should fix them!

**Comment by hameerabbasi** [comment]
Thanks, @joaosferreira!

**Pull request merged**